x/harvest/keeper/borrow.go

```diff
@@ -0,0 +1,138 @@
+package keeper
+
+import (
+	"math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/kava-labs/kava/x/harvest/types"
+)
+
+// USDX is the USDX coin's denom
+const USDX = "usdx"
+
+// Borrow funds
+func (k Keeper) Borrow(ctx sdk.Context, borrower sdk.AccAddress, amount sdk.Coin) error {
+
+	err := k.ValidateBorrow(ctx, borrower, amount)
+	if err != nil {
+		return err
+	}
+
+	err = k.supplyKeeper.SendCoinsFromAccountToModule(ctx, borrower, types.ModuleAccountName, sdk.NewCoins(amount))
+	if err != nil {
+		return err
+	}
+
+	borrow, found := k.GetBorrow(ctx, borrower, amount.Denom)
+	if !found {
+		borrow = types.NewBorrow(borrower, amount)
+	} else {
+		borrow.Amount = borrow.Amount.Add(amount)
+	}
+
+	k.SetBorrow(ctx, borrow)
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeHarvestBorrow,
+			sdk.NewAttribute(sdk.AttributeKeyAmount, amount.String()),
+			sdk.NewAttribute(types.AttributeKeyBorrower, borrow.Borrower.String()),
+			sdk.NewAttribute(types.AttributeKeyBorrowDenom, borrow.Amount.Denom),
+		),
+	)
+
+	return nil
+}
+
+// ValidateBorrow validates a borrow request against borrower and protocol requirements
+func (k Keeper) ValidateBorrow(ctx sdk.Context, borrower sdk.AccAddress, amount sdk.Coin) error {
+	if err := k.validateBorrowUser(ctx, borrower, amount); err != nil {
+		return err
+	}
+
+	if err := k.validateBorrowSystem(ctx, borrower, amount); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k Keeper) validateBorrowUser(ctx sdk.Context, borrower sdk.AccAddress, amount sdk.Coin) error {
+	var borrowMoneyMarket types.MoneyMarket
+	var borrowAssetPrice sdk.Dec
+	var borrowUSDValue sdk.Dec
+	if amount.Denom == USDX {
+		borrowUSDValue = sdk.NewDecFromInt(amount.Amount).Quo(sdk.NewDec(int64(math.Pow(10, 6)))) // TODO: include conversion_factor in params
+	} else {
+		locBorrowMoneyMarket, found := k.GetMoneyMarket(ctx, amount.Denom)
+		if !found {
+			return sdkerrors.Wrapf(types.ErrMarketNotFound, "no market found for denom %s", amount.Denom)
+		}
+		locBorrowAssetPrice, err := k.pricefeedKeeper.GetCurrentPrice(ctx, locBorrowMoneyMarket.SpotMarketID)
+		if err != nil {
+			return sdkerrors.Wrapf(types.ErrPriceNotFound, "1: no price found for market %s", locBorrowMoneyMarket.SpotMarketID)
+		}
+		borrowMoneyMarket = locBorrowMoneyMarket
+		borrowAssetPrice = locBorrowAssetPrice.Price
+		borrowUSDValue = sdk.NewDecFromInt(amount.Amount).Mul(borrowAssetPrice)
+	}
+
+	// Get the user's deposits
```

Comment:
move the `len(deposits) == 0` check to before checking USD value of the borrow.

Reply:
Moved as requested in https://github.com/Kava-Labs/kava/pull/702/commits/5428519008577591be427331d9501c142a1e77c8.

Argument for moving it back up: if the user has no deposits then it'll be impossible to successfully borrow an asset. So shouldn't we throw this error once as soon as we know the user has no deposits?

Reply:
Yeah, that's what I meant. This should be the first check and we can short-circuit error if the user has no deposits. Sorry if unclear.